### ISSUE TITLE
removable drive option does not work

### ISSUE THREAD

**Issue by the-Arioch** [DDoSolitary/yasfw#28]
Hello! thanks for the program. I (a Windows guy) have to mantain an isolated CentOS 6.0 box, and it was minimal (no NFS, no SMB, no modern git no any git TUI/GUI), so without an SSH FS i'd be pretty downed. Guess it was a nive way for you to learn Rust :-)

Now to the meat of it.

```
echo {password}| yasfw -r -c --server {IP} --user root --mount-point W: -C /vz/private/{something}

Sep 23 14:58:07.066 INFO initializing, dokan_driver_version: 400, dokan_version:
 150
Sep 23 14:58:07.139 INFO connection established, server_public_key: CF:23:........:30:A0
Password:
Sep 23 14:58:07.236 INFO password authentication succeeded
Sep 23 14:58:07.498 INFO mounted
```

Now if i try to eject the drive using usual GUI (Total Commander in this case, but it would use Explorer's ActiveX context menu anyway) i get error:

```
[Window Title]
Microsoft Windows
[Content]
Ошибка при извлечении "root@193.169.173.82:22 (W:)".
[ОК]
```

I use Win7 sp1 pro x64 rus, the error just says "ejection error" without details.

Also

```
d:\SSH>dokanctl /u 0
Driver path: 'C:\Windows\system32\drivers\dokan1.sys'
send global release for 0:
DokanError: Ioctl 0x92010 failed with code 122 on Device \\.\Dokan_1
Failed to unmount: 0:
Unmount status = 1
```

YASFW does not display the remote-drive option warning (maybe it does not have to, but franky i think for beta software you better be verbose when you can), so i am not sure it was even attempted, anyway, it does not work and i can only exit by Ctrl+C

**Comment by DDoSolitary** [comment]
The removable functionality is provided by Dokan and I simply passed a flag to enable it. Please try their "mirror" (a sample program that simply mirrors an existing directory) and I will investigate if the error is unique to yasfw.

**Comment by the-Arioch** [comment]
i am total newb to all this

do you mean i have to mount an existign NTFS folder as a ghost drive (using that `mirror` just like old skool `subst` command) and try to eject that drive then?

**Comment by DDoSolitary** [comment]
Yes, exactly.

**Comment by the-Arioch** [comment]
```
cd /d c:\Program Files\Dokan\Dokan Library-1.5.0\sample\mirror\

c:\Program Files\Dokan\Dokan Library-1.5.0\sample\mirror>mirror /r d:\books  /l
P: /m
Success
```

Did i miss somethign or should had this worked?

It seems it fails umounting the same way.

What can be done to get more useful logs from Dokan?

**Comment by the-Arioch** [comment]
Still would be better if yasfw acknowledged that option publicly, so would be somewhat less uncertainty who is most probably in the wrong, yasfw or dokan

**Comment by DDoSolitary** [comment]
Dokan is a kernel driver and I don't think there's anything that can be done from user-mode side. I would suggest open an issue in Dokan's repo. Their issue template and wiki also includes instructions to get various log data.

**Comment by DDoSolitary** [comment]
> YASFW does not display the remote-drive option warning (maybe it does not have to, but franky i think for beta software you better be verbose when you can), so i am not sure it was even attempted, anyway, it does not work and i can only exit by Ctrl+C

I guess you mean the removable drive options? I don't think it's very necessary to print that information during normal operation. I'm not sure but it's likely that Windows provides some way to check if a drive is removable. And for debugging purposes, it's always possible to get some clues from Dokan's log: https://github.com/dokan-dev/dokany/blob/4208150968d650cc4102c07b715b4c9658c42afb/sys/event.c#L725

**Comment by the-Arioch** [comment]
Just skimmed through https://github.com/dokan-dev/dokany/wiki/How-to-Debug-Dokan and makign debug installation, and on a workign machine, a bit freaks me out...

Interest thing is `mirror` help output

```
  /n (use network drive)                         Show device as network device.
  /m (use removable drive)                       Show device as removable media.
...
  /x Network unmount                             Allows unmounting network drive from file explorer.
```

Seems that "allows" and "shows" are not bound... At least for network drives.

`mirror ... /n /x` works (albeit the menu then is Disconnect not Eject
`mirror ... /m /x` expectedly does not...

**Comment by DDoSolitary** [comment]
Network drives are separate things. By default, Dokan mounts as an ordinary drive, which provides more features but requires the presence of the kernel driver, while mounting as network drives is also supported as an alternative. Unmounting is expected to work with the removable option enabled along. The /x option corresponds to the DOKAN_OPTION_ENABLE_UNMOUNT_NETWORK_DRIVE and is only meaningful for network drives. The flag is relatively new and has not been introduced when I wrote the option handling code for yasfw, so it's not supported by yasfw for now.

**Comment by the-Arioch** [comment]
```
Unmounting
A file system can be unmounted by calling the function DokanUnmount or DokanRemoveMountPoint. In most cases when the programs or shells using the file system hang, the unmount operation will solve the problem by bringing the system to a previous state when the file system was not mounted.

Users can use the command line to unmount file system like this:

dokanctl.exe /u DriveLetter
```

https://dokan-dev.github.io/dokany-doc/html/

So, it seems there SHOULD be some hooks in the core, but probably they expect somethign from yasfw too, some callback?

Or maybe not, if `mirror` does not work too...

**Comment by DDoSolitary** [comment]
This is for unmounting programmatically. I don't think there's anything additional needs to be done from yasfw side.

**Comment by the-Arioch** [comment]
https://groups.google.com/g/dokan/c/maWhQ-LInVk/m/aXrd-phGyHsJ
Suggests that DOKAN_OPERATIONS.Unmount( ) callback has to be called.

https://groups.google.com/g/dokan/c/esYVLIBgfoM/m/mdt03Zor26cJ
This suggests it should be applicatio nerror not core error, however ti is truly oooold (dokan 0.6.0)

And then.... a fatality?
https://groups.google.com/g/dokan/c/1GPwYfSUuFs/m/_Hg0kd8zAwAJ

> Eject is not supported so it is normal but what is the output of dokanctl /u M ?
> Liryna   18 SEP 2016 
> about FUSE

Hmm, at least it seems it is two different issues - docanctl /u and Windows Explorer's eject.

**Comment by the-Arioch** [comment]
> This is for unmounting programmatically

yep, and it should be invoked by docanctl /u - and it fails to me.

let check it the other side, can YOU on your box mount some SSH as a removable drive and then try to eject it both from Explorer and from dokanctl? would it work on your system or not?

**Comment by the-Arioch** [comment]
Daaarn!!!! is it really so stupid???  wait a minute

**Comment by Liryna** [comment]
Did you try `dokanctl /u W` ?

**Comment by the-Arioch** [comment]
Seems unmount-by-number got broken! do you use latest released Dokan ( .3000 ) or some newer dev-build ?

```
c:\Program Files\Dokan\Dokan Library-1.5.0>dokanctl /?
Driver path: 'C:\Windows\system32\drivers\dokan1.sys'
dokanctl /u MountPoint
dokanctl /u M
dokanctl /i [d||n]
dokanctl /r [d||n]
dokanctl /v

Example:
  /u M                : Unmount M: drive
  /u C:\mount\dokan   : Unmount mount point C:\mount\dokan
  /u 1                : Unmount mount point 1
  /i d                : Install driver
  /i n                : Install network provider
  /r d                : Remove driver
  /r n                : Remove network provider
  /l a                : List current mount points
  /d [0-7]            : Enable Kernel Debug output
  /v                  : Print Dokan version
```

Seems straightforward.

```
c:\Program Files\Dokan\Dokan Library-1.5.0>dokanctl /l
Driver path: 'C:\Windows\system32\drivers\dokan1.sys'
  Mount points: 2
  0# MountPoint: \DosDevices\P: - UNC:  - DeviceName: \Device\Volume{d6cc17c5-17
28-4085-bce7-964f1e9f5de9}
  1# MountPoint: \DosDevices\W: - UNC:  - DeviceName: \Device\Volume{d6cc17c5-17
29-4085-bce7-964f1e9f5de9}
```

First is mirror, second is yafsw.

```
c:\Program Files\Dokan\Dokan Library-1.5.0>dokanctl /u 0
Driver path: 'C:\Windows\system32\drivers\dokan1.sys'
send global release for 0:
DokanError: Ioctl 0x92010 failed with code 122 on Device \\.\Dokan_1
Failed to unmount: 0:
Unmount status = 1

c:\Program Files\Dokan\Dokan Library-1.5.0>dokanctl /u 1
Driver path: 'C:\Windows\system32\drivers\dokan1.sys'
send global release for 1:
DokanError: Ioctl 0x92010 failed with code 122 on Device \\.\Dokan_1
Failed to unmount: 1:
Unmount status = 1
```

Ooooops! But here were were from the get go.

```
c:\Program Files\Dokan\Dokan Library-1.5.0>dokanctl /u P
Driver path: 'C:\Windows\system32\drivers\dokan1.sys'
send global release for P:
Unmount status = 0

c:\Program Files\Dokan\Dokan Library-1.5.0>dokanctl /u W
Driver path: 'C:\Windows\system32\drivers\dokan1.sys'
send global release for W:
Unmount status = 0
```

Wooops!

Talk about looking for black cat in black night....

**Comment by DDoSolitary** [comment]
> Suggests that DOKAN_OPERATIONS.Unmount( ) callback has to be called.

It's already implemented. Otherwise even normal unmounting without the removable option won't work properly. https://github.com/DDoSolitary/yasfw/blob/91c0a68654432e0d9760658ad737973b99a484d8/src/main.rs#L785

**Comment by the-Arioch** [comment]
@Liryna okay, since you luckily dropped in here, can you tell me, what is problematic in forwarding Windows Explorer Eject into `dokanctl /u` ? one thing i saw potentially blocking was - UAH (but frankly, if Eject menu item is displayed for a user, then that user would be given some option, sudo-like probably, to eject hardware disk).

This functionality seems to be mantioned in maillist since 2016 if not 2011, and it seems quite a basic one, there should be realyl some heavy block in the road?

**Comment by the-Arioch** [comment]
@Liryna also, w.r.t. `  /u 1                : Unmount mount point 1`

While i never saw it "in the wild", but there were talks that API-wise WinNT allows all 255 not-null ASCII characters to act as drive letters. And someone got tripped over it, actually making such a drive "letter" and then being unable to do anything with it short of reboot. Then they started discussing using non-Latin letters as drive letters, etc. Just some loose memory here...

However, if that actually IS possible API-wise, then it seems to be ambiguity whether `/u 1` would mean topmost mountpoint, second mountpoitn ("1#") - or a (probably non-existent) drive letter `1:\`

That said, i never saw such a thing in the wild, so perhaps no practical problem in that.

**Comment by the-Arioch** [comment]
Seems Dokan documentation problem, the `/u 1                : Unmount mount point 1` example seems no more implemented.

Not `yafsw` issue.
Sorry for trouble.

**Comment by Liryna** [comment]
`IOCTL_STORAGE_EJECT_MEDIA` should be supported
https://github.com/dokan-dev/dokany/blob/8ca918dddb960ad0af17eaa1984b0e5dbf007315/sys/device.c#L523-L525
but that's only for removable drive
https://docs.microsoft.com/en-us/windows/win32/api/winioctl/ni-winioctl-ioctl_storage_eject_media
So `mirror.exe /r C:\Users /l m /m` with the `/m` for removable drive type should make it work.

I must say `/u 1 : Unmount mount point 1` is probably legacy :( I never used it.

**Comment by the-Arioch** [comment]
> should make it work.

Not for me...  https://github.com/DDoSolitary/yasfw/issues/28#issuecomment-925768670

Granted, i run `mirror` from elevated console, just like dokanctl. Any issue with being elevated?..

```
c:\Program Files\Dokan\Dokan Library-1.5.0\sample\mirror>mirror /r d:\books  /l
p: /m
[Mirror] Failed to add security privilege to process
        => GetFileSecurity/SetFileSecurity may not work properly
        => Please restart mirror sample with administrator rights to fix it
```

So, without elevation it does not work, and with elevation it does not eject.
Also, your 2016 mail (albeit about FUSE) suggested Windows Explorer menu's eject was not implemented in Dokan (maybe was implemented since then)

**Comment by the-Arioch** [comment]
"Curiousier and curiousier"

Using "/o" and not-elevated.

```
c:\Program Files\Dokan\Dokan Library-1.5.0\sample\mirror>mirror /r d:\books  /l
p: /m /o
[Mirror] Failed to add security privilege to process
        => GetFileSecurity/SetFileSecurity may not work properly
        => Please restart mirror sample with administrator rights to fix it
Success
```

Elevated `c:\>mountvol p: /d` DID unmount the `mirror` however did hung after that, it does not return to command prompt.

And the whole computer seems going sinking.... Can not open Explorer at all

**Comment by the-Arioch** [comment]
It is dead.... gonna for hard reboot.

**Comment by the-Arioch** [comment]
Back online for awhile.
I won't say this Win7 sp1 pro box is most clean in the world, it is not.
Yet Dokan seems to act out on it.

I hope @Liryna  would see it more like "testcase" than deviation and would help me to investigate and hopefulyl fix (or workaround) this.

Ejecting the "removable" mirror (or yasfw) drive here does not work. And it seems rather hard to determine why, where iexecution flow runs astray.

**Comment by the-Arioch** [comment]
> DokanError: Ioctl 0x92010 failed with code 122 on Device \\.\Dokan_1

* https://social.technet.microsoft.com/wiki/contents/articles/24653.decoding-io-control-codes-ioctl-fsctl-and-deviceiocodes-with-table-of-known-values.aspx
* http://www.osronline.com/article.cfm%5Earticle=229.htm

0x92010  ==> FILE_SYSTEM (0x9) // function 0x804 // FILE_ANY_ACCESS // METHOD_BUFFERED

What is that 0x0804 ? IOCTL_STORAGE_EJECT_MEDIA is function 0x0202

**Comment by the-Arioch** [comment]
@Liryna  i took a "conventional wisdom" snippet from 

https://stackoverflow.com/questions/58735900/how-to-eject-usb-drive-on-windows-10-ioctl-storage-eject-media-no-longer-enough

```
CreateFile (drive letter, with read/write rights, file share read and write)
DeviceIoControl(FSCTL_LOCK_VOLUME) 
DeviceIoControl(FSCTL_DISMOUNT_VOLUME) 
DeviceIoControl(IOCTL_STORAGE_MEDIA_REMOVAL) PreventMediaRemoval = FALSE 
DeviceIoControl(IOCTL_STORAGE_EJECT_MEDIA)
```

Notice, they said it ceased working on modern Windows and in answer they suggest another function.

If it matters, a readymade implementation from https://forum.lazarus.freepascal.org/index.php?topic=34168.0

OK:   CreateFile   -- VolumeName '\\.\p:', GENERIC_READ or GENERIC_WRITE,
OK:   FSCTL_LOCK_VOLUME
ERROR:  FSCTL_DISMOUNT_VOLUME  -- GetLastError = 1, allegedly "wrong function" or something.
NOT tried:  IOCTL_STORAGE_MEDIA_REMOVAL
NOT tried:  IOCTL_STORAGE_EJECT_MEDIA

The same on a DVD-ROM drive

OK:   CreateFile   -- VolumeName '\\.\r:', GENERIC_READ
OK:   FSCTL_LOCK_VOLUME
OK:  FSCTL_DISMOUNT_VOLUME 
OK:  IOCTL_STORAGE_MEDIA_REMOVAL
OK:  IOCTL_STORAGE_EJECT_MEDIA

....and tray is out

Now the trick, i override and ignore FSCTL_DISMOUNT_VOLUME for dokany/mirror disk - and now i suddenly see it exactly as above with `mountvol` - on IOCTL_STORAGE_EJECT_MEDIA call the mirror.exe exits, but... the call hangs and never returns.

**Comment by the-Arioch** [comment]
Main thread (ThreadID 6712):

````
ntoskrnl.exe!KiCpuId+0xaa
ntoskrnl.exe!ObfReferenceObject+0x792
ntoskrnl.exe!KeWaitForMutexObject+0x1a3
ntoskrnl.exe!IoReleaseRemoveLockAndWaitEx+0x45
dokan1.sys+0x1285f
dokan1.sys+0x15dd0
dokan1.sys+0x3bc0
dokan1.sys+0x46f7
dokan1.sys+0x4f30
dokan1.sys+0x48ac
dokan1.sys+0x6888
dokan1.sys+0x5efe
fltmgr.sys!FltIsCallbackDataDirty+0x2df
fltmgr.sys+0x16df
ntoskrnl.exe!CcCopyRead+0x4aa
ntoskrnl.exe!NtQueryVolumeInformationFile+0x1251
ntoskrnl.exe!NtDeviceIoControlFile+0x56
ntoskrnl.exe!longjmp+0x5bf3
wow64cpu.dll!TurboDispatchJumpAddressEnd+0x6c0
wow64cpu.dll!TurboDispatchJumpAddressEnd+0x1fb
wow64.dll!Wow64SystemServiceEx+0x1ce
wow64.dll!Wow64KiUserCallbackDispatcher+0x204
wow64win.dll+0x32bf9
ntdll.dll!KiUserCallbackDispatcher+0x1f
wow64win.dll+0x3fdaa
wow64win.dll+0x1aee6
wow64win.dll+0x327ef
wow64win.dll+0x1b022
wow64.dll!Wow64SystemServiceEx+0xd7
wow64cpu.dll!TurboDispatchJumpAddressEnd+0x2d
wow64.dll!Wow64SystemServiceEx+0x1ce
wow64.dll!Wow64LdrpInitialize+0x42a
ntdll.dll!RtlImageDirectoryEntryToData+0x19a3
ntdll.dll!longjmp+0x2caa0
ntdll.dll!LdrInitializeThunk+0xe
ntdll.dll!NtDeviceIoControlFile+0x15
kernel32.dll!DeviceIoControl+0x48
project1.exe+0x111655
project1.exe+0x1116cc
project1.exe+0x11172f
project1.exe+0x86665
project1.exe+0x8aa6c
project1.exe+0x9cad9
project1.exe+0x8abbc
project1.exe+0x8aa6c
project1.exe+0xffe1e
project1.exe+0x8a0bf
project1.exe+0x5321e
USER32.dll!gapfnScSendMessage+0x33a
USER32.dll!GetThreadDesktop+0xd7
USER32.dll!GetThreadDesktop+0x185
USER32.dll!GetThreadDesktop+0x1e1
ntdll.dll!KiUserCallbackDispatcher+0x2e
USER32.dll!SendMessageW+0x4c
USER32.dll!gapfnScSendMessage+0x33a
USER32.dll!GetThreadDesktop+0xd7
USER32.dll!GetClientRect+0xc5
USER32.dll!CallWindowProcW+0x1b
project1.exe+0x8ab69
project1.exe+0x8aa6c
project1.exe+0x9cad9
project1.exe+0x5321e
USER32.dll!gapfnScSendMessage+0x33a
USER32.dll!GetThreadDesktop+0xd7
USER32.dll!CharPrevW+0x138
USER32.dll!DispatchMessageW+0xf
project1.exe+0x109010
project1.exe+0x109053
project1.exe+0x109392
project1.exe+0x11918c
kernel32.dll!BaseThreadInitThunk+0x12
ntdll.dll!RtlInitializeExceptionChain+0x63
ntdll.dll!RtlInitializeExceptionChain+0x36
````

aux thread old (ThreadID 3736)

```
ntoskrnl.exe!KiCpuId+0xaa
ntoskrnl.exe!ObfReferenceObject+0x792
ntoskrnl.exe!KeWaitForMutexObject+0x1a3
ntoskrnl.exe!CcGetFlushedValidData+0xc34
ntoskrnl.exe!KeInsertQueueApc+0x605
ntoskrnl.exe!ObfReferenceObject+0x99d
ntoskrnl.exe!KeWaitForMultipleObjects+0x272
ntoskrnl.exe!NtRequestWaitReplyPort+0x335
ntoskrnl.exe!FsRtlOplockFsctrlEx+0x1a08
ntoskrnl.exe!longjmp+0x5bf3
wow64cpu.dll!TurboDispatchJumpAddressEnd+0x6c0
wow64cpu.dll!TurboDispatchJumpAddressEnd+0xf5
wow64.dll!Wow64SystemServiceEx+0x1ce
wow64.dll!Wow64LdrpInitialize+0x42a
ntdll.dll!longjmp+0x2cbe4
ntdll.dll!LdrInitializeThunk+0xe
ntdll.dll!NtWaitForMultipleObjects+0x15
kernel32.dll!BaseThreadInitThunk+0x12
ntdll.dll!RtlInitializeExceptionChain+0x63
ntdll.dll!RtlInitializeExceptionChain+0x36
```

aux thread new (threadID 6712)
```
ntoskrnl.exe!KiCpuId+0xaa
ntoskrnl.exe!ObfReferenceObject+0x792
ntoskrnl.exe!KeWaitForMutexObject+0x1a3
ntoskrnl.exe!CcGetFlushedValidData+0xc34
ntoskrnl.exe!KeInsertQueueApc+0x605
ntoskrnl.exe!ObfReferenceObject+0x99d
ntoskrnl.exe!KeRemoveQueueEx+0x323
ntoskrnl.exe!PsAssignImpersonationToken+0x13c
ntoskrnl.exe!KeReleaseSemaphore+0x1095
ntoskrnl.exe!longjmp+0x5bf3
ntdll.dll!NtWaitForWorkViaWorkerFactory+0xa
wow64.dll!Wow64EmulateAtlThunk+0x1c9df
wow64.dll!Wow64SystemServiceEx+0xd7
wow64cpu.dll!TurboDispatchJumpAddressEnd+0x2d
wow64.dll!Wow64SystemServiceEx+0x1ce
wow64.dll!Wow64LdrpInitialize+0x42a
ntdll.dll!longjmp+0x2cbe4
ntdll.dll!LdrInitializeThunk+0xe
ntdll.dll!ZwWaitForWorkViaWorkerFactory+0x12
kernel32.dll!BaseThreadInitThunk+0x12
ntdll.dll!RtlInitializeExceptionChain+0x63
ntdll.dll!RtlInitializeExceptionChain+0x36
```

**Comment by Liryna** [comment]
Thanks for looking into this!

Since here you are starting to look into the kernel communication, I would suggest to build a début version of the driver and enable the debug logs. You will see the activity in it.
Also, procmon is a good software to find out what is the call hanging by filter with process name.

**Comment by the-Arioch** [comment]
Threads are unkillable, so i am going sink again. But okay, have to go home anyway :-)

I wish there was somethign to trace api for Windows.  Easy "point and click" detours toolkits actually did not work for me even for as simple thing as GetsystemMetrix. And how to custom made one for this case i dunno.

Last check, `mirror` without `/o /e /f` - it seems something is shakey about mounts manager in either my Win7 or in dokan or in yasfw (in this latter case - still in dokan too).

....no good, i still got stuck. `mirror` quit ok, but the call to `IOCTL_STORAGE_EJECT_MEDIA` did not get out.

This time early on i even had FIVE threads but three were (as shown by Process Explorer)
3104			ntdll.dll!TpCallbackIndependent+0x22b
Of those soon only one remained.

But the remaining three threads are unkillable and do not return either.

**Comment by the-Arioch** [comment]
```
0	ntoskrnl.exe!KiCpuId+0xaa
1	ntoskrnl.exe!ObfReferenceObject+0x792
2	ntoskrnl.exe!KeWaitForMutexObject+0x1a3
3	ntoskrnl.exe!IoReleaseRemoveLockAndWaitEx+0x45
4	dokan1.sys+0x1285f
5	dokan1.sys+0x15dd0
6	dokan1.sys+0x3bc0
7	dokan1.sys+0x46f7
8	dokan1.sys+0x4f30
9	dokan1.sys+0x48ac
10	dokan1.sys+0x6888
11	dokan1.sys+0x5efe
12	fltmgr.sys!FltIsCallbackDataDirty+0x2df
13	fltmgr.sys+0x16df
14	ntoskrnl.exe!CcCopyRead+0x4aa
15	ntoskrnl.exe!NtQueryVolumeInformationFile+0x1251
16	ntoskrnl.exe!NtDeviceIoControlFile+0x56
17	ntoskrnl.exe!longjmp+0x5bf3
18	wow64cpu.dll!TurboDispatchJumpAddressEnd+0x6c0
19	wow64cpu.dll!TurboDispatchJumpAddressEnd+0x1fb
20	wow64.dll!Wow64SystemServiceEx+0x1ce
21	wow64.dll!Wow64KiUserCallbackDispatcher+0x204
22	wow64win.dll+0x32bf9
23	ntdll.dll!KiUserCallbackDispatcher+0x1f
24	wow64win.dll+0x3fdaa
25	wow64win.dll+0x1aee6
26	wow64win.dll+0x327ef
27	wow64win.dll+0x1b022
28	wow64.dll!Wow64SystemServiceEx+0xd7
29	wow64cpu.dll!TurboDispatchJumpAddressEnd+0x2d
30	wow64.dll!Wow64SystemServiceEx+0x1ce
31	wow64.dll!Wow64LdrpInitialize+0x42a
32	ntdll.dll!RtlImageDirectoryEntryToData+0x19a3
33	ntdll.dll!longjmp+0x2caa0
34	ntdll.dll!LdrInitializeThunk+0xe
35	ntdll.dll!NtDeviceIoControlFile+0x15
36	kernel32.dll!DeviceIoControl+0x48
....the rest is mostly GUI framework
```

At this point i am lost and don't know what to do.

**Comment by the-Arioch** [comment]
You can (very and VERY unwisely) try this, just to score few extra lazyness points :-)

it allows you to override IOCTL failures

[eject-test-usb.zip](https://github.com/DDoSolitary/yasfw/files/7219622/eject-test-usb.zip)

**Comment by the-Arioch** [comment]
@Liryna 
> procmon is a good software to find out what is the call hanging by filter with process name.

i tried to intercept Explorer's Eject menu with it - nothing of interest.
i think ProcMon uses userland hooks and does not drill down, like ooold DiskMon did.
or maybe i don't know how to configure it to do so.

> to build a début version of the driver and enable the debug logs

Like i said in another issue, a perspective to do it on my main work machine freaks me out.
Also, i neither have VC++ nor any experience with it :-)

However you probably can provide PDB files for release driver?

**Comment by the-Arioch** [comment]
Anyway grep shows there is the only call to `IoReleaseRemoveLockAndWait(&dcb->RemoveLock, RequestContext);` in dokan sources and it is near the exit of `NTSTATUS DokanEventRelease`

And frankly debug logs would give me but nothign here, as it is about multi-thread locks counting and miscounting, and while this function _probably_ correctly waits for someone to decrement lock counter - it is that misbehaving or missed someone to blame. How would a log show it to me?..

**Comment by the-Arioch** [comment]
https://github.com/microsoft/Windows-driver-samples/blob/7895dd22785ddba5e973662ed942be3b3452b89d/general/event/wdm/event.c#L361

Notice how ReleaseAndWait here is called IMMEDIATELY after acquire, and all the heavy (and slow) lifting only happend after it.

It makes some sense as the only thing you seem to be in need there is raising a barrier and waiting everyone else to jump over it.

Before you called ReleaseAndWait (that is, IN BETWEEN calls to acquire and release) there still can be other threads doing random jobs. Like closing file handles, dunno.

But since you already are doing things like shutting down even loops, etc, even BEFORE you ensured no other work is pending by call to ReleaseAndWait, there perhaps MAY be a chance, that some worker thread getting confused and never able to release the lock.

Just speculating.
